ui/src/app/routes.tsx

```diff
@@ -0,0 +1,265 @@
+    render: () => <WorkspaceWrapper><PageLayout><DetailPage /></PageLayout></WorkspaceWrapper>,
+    title: 'Review Cohort Participants',
+    breadcrumb: BreadcrumbType.Participant
+  },
+  {
+    name: 'concepts',
+    path: '/workspaces/:ns/:wsid/concepts',
+    render: () => <WorkspaceWrapper><ConceptHomepage /></WorkspaceWrapper>,
+    title: 'Search Concepts',
+    breadcrumb: BreadcrumbType.Workspace
+  },
+  {
+    name: 'data',
+    path: '/workspaces/:ns/:wsid/data',
+    render: () => <WorkspaceWrapper><DataPage /></WorkspaceWrapper>,
+    title: 'Data Page',
+    breadcrumb: BreadcrumbType.Workspace
+  },
+  {
+    name: 'datasets',
+    path: '/workspaces/:ns/:wsid/data/datasets',
+    render: () => <WorkspaceWrapper><DataSet /></WorkspaceWrapper>,
+    title: 'Dataset Page',
+    breadcrumb: BreadcrumbType.Dataset
+  },
+  {
+    name: 'conceptSets',
+    path: '/workspaces/:ns/:wsid/concepts/sets',
+    render: () => <WorkspaceWrapper><ConceptSetList /></WorkspaceWrapper>,
+    title: 'View Concept Sets',
+    breadcrumb: BreadcrumbType.Workspace
+  },
+  {
+    name: 'conceptSet',
+    path: '/workspaces/:ns/:wsid/concepts/sets/:csid',
+    render: () => <WorkspaceWrapper><ConceptSetDetails /></WorkspaceWrapper>,
+    title: 'Concept Set',
+    breadcrumb: BreadcrumbType.ConceptSet
+  },
+  {
+    name: 'workspaceReview',
+    path: '/admin/review-workspace',
+    render: () => <AdminReviewWorkspace />,
+    title: 'Review Workspaces'
+  },
+  {
+    name: 'users',
+    path: '/admin/user',
+    render: () => <AdminUser />,
+    title: 'User Admin Table'
+  },
+  {
+    name: 'profile',
+    path: '/profile',
+    render: () => <ProfilePage />,
+    title: 'Profile'
+  },
+  {
+    name: 'settings',
+    path: '/settings',
+    render: () => <Settings />,
+    title: 'Settings'
+  },
+  {
+    name: 'workspaceCreate',
+    path: '/workspaces/build',
+    render: () => <WorkspaceEdit mode={WorkspaceEditMode.Create} />,
+    title: 'Create Workspace'
+  },
+];
+
+const handlers = routes.map(({path, ...rest}) => {
+  const keys = []; // mutated by pathToRegexp
+  const regex = pathToRegexp(path, keys);
+  return {
+    regex, ...rest,
+    keys: keys.map(k => k.name),
+    makePath: pathToRegexp.compile(path)
+  };
+});
+
+const handlersByName = fp.keyBy('name', handlers);
+
+const parseRoute = ({pathname, search}) => {
+  const handler = handlers.find(({regex}) => regex.test(pathname));
+  return handler && {
+    ...handler,
+    params: fp.zipObject(handler.keys, fp.tail(handler.regex.exec(pathname))),
+    query: qs.parse(search, {ignoreQueryPrefix: true, plainObjects: true}),
+  };
+};
+
+/**
+ * HOC that injects a `route` prop, whose value is the matched route handler, extended
+ * with `params` (path params) and `query` (query params). If no route matches, the value
+ * will be undefined.
+ */
+export const withRoute = () => WrappedComponent => {
```

Comment:
This is the main tool for injecting and using information about the current route.

An important and useful detail here is that the route _matching_ is separate from the route _rendering_. Any component can use this HOC to make decisions based on the route, even ones higher in the tree than the `Router` component itself.

For example, we will likely have a top-level 'signed in' wrapper that will examine the route to determine whether to redirect to the login page. There could also be a 'title renderer' component whose sole task is to watch the current route and update the page title. The `Router` component is just one of many consumers of this data, which can all be freely composed with one another.